### PR TITLE
Stop using uuid for request_id

### PR DESCRIPTION
**Pull request by pipermerriam**
Builds on #99 

## What was wrong?

The cost of generating a UUID is *maybe* more expensive.

## How was it fixed?

Custom scheme that *might* have minimal overhead.

- [x] TODO: benchmark suite does not flex the `request/response` API.

#### Cute Animal Picture

35843c422aaa96acff54869fa97ef8e4

### CODE REVIEW

**Comment by pipermerriam** [comment]
Benchmarking suggest this is indeed measurably faster.  Cuts total time generating request id in the `request` method from being around `1.3%` to be `0.2%`.

File: lahja/asyncio/endpoint.py [hunk 1]
```diff
@@ -277,7 +279,7 @@ class AsyncioEndpoint(BaseEndpoint):
     _receiving_queue: "asyncio.Queue[Tuple[Union[bytes, BaseEvent], Optional[BroadcastConfig]]]"
     _receiving_loop_running: asyncio.Event
 
-    _futures: Dict[Optional[str], "asyncio.Future[BaseEvent]"]
```
Comment (lithp) [inline lahja/asyncio/endpoint.py]: I'm surprised it was even possible for the key of a dictionary to be optional :smile:

File: lahja/asyncio/endpoint.py [hunk 2]
```diff
@@ -286,17 +288,25 @@ class AsyncioEndpoint(BaseEndpoint):
     _sync_handler: DefaultDict[Type[BaseEvent], List[SubscriptionSyncHandler]]
 
     _loop: Optional[asyncio.AbstractEventLoop] = None
+    _get_request_id: Iterator[RequestID]
 
     def __init__(self, name: str) -> None:
         self.name = name
 
+        try:
+            self._get_request_id = RequestIDGenerator(name.encode("ascii") + b":")
```
Comment (lithp) [inline lahja/asyncio/endpoint.py]: nitpick: In an imaginary future where there are multiple backends users might want to switch between different kinds of endpoints. If this check is on the Asyncio endpoint but not the Trio one then non-ascii users could have a hard time experimenting with Asyncio after building their app with Trio.

Putting this check into `BaseEndpoint` would make life easier for those hypothetical users.

File: lahja/asyncio/endpoint.py [hunk 3]
```diff
@@ -734,7 +747,9 @@ def broadcast_nowait(
 
         return result
 
-    def _remove_cancelled_future(self, id: str, future: "asyncio.Future[Any]") -> None:
+    def _remove_cancelled_future(
+        self, id: RequestID, future: "asyncio.Future[Any]"
+    ) -> None:
```
Comment (lithp) [inline lahja/asyncio/endpoint.py]: nitpick: Is this Black? I think our convention has been to not add this final newline

File: lahja/base.py [hunk 4]
```diff
@@ -219,6 +219,12 @@ class BaseEndpoint(EndpointAPI):
 
     logger = logging.getLogger("lahja.endpoint.Endpoint")
 
+    def __str__(self) -> str:
+        return f"Endpoint[{self.name}]"
+
+    def __repr__(self) -> str:
+        return f"<{self.name}>"
```
Comment (lithp) [inline lahja/base.py]: nitpick: I think these two are flipped, [`__repr__`](https://docs.python.org/3/reference/datamodel.html#object.__repr__) is the one which roughly looks like a constructor. What's the rationale for implementing both? I think `__str__` is usually added if `__repr__` gives output which is too long, but here they both give short and printable output.

File: lahja/asyncio/endpoint.py (hunk 1 above)
Comment (pipermerriam) [reply lahja/asyncio/endpoint.py]: Just mean `None` as a key which is possible, but you can only ever have one.

File: lahja/asyncio/endpoint.py (hunk 2 above)
Comment (pipermerriam) [reply lahja/asyncio/endpoint.py]: :+1:

File: lahja/asyncio/endpoint.py (hunk 3 above)
Comment (pipermerriam) [reply lahja/asyncio/endpoint.py]: black

File: lahja/asyncio/endpoint.py (hunk 1 above)
Comment (cburgdorf) [reply lahja/asyncio/endpoint.py]: This is one of the few occasions where `Union[str, None]` (which is equivalent to `Optional[str]`) may *feel* a little more suitable :sweat_smile:

**Pull request merged**